KAFKA-17411: Add new StateStore interfaces

nicktelford opened the pull request:
These are the new interfaces detailed in KIP-1035: "StateStore managed
changelog offsets".

This PR introduces the interfaces changes, but makes otherwise no
consequential behavioural changes.

Outside of `StateStore.java`, _all_ changes are essentially just a
rename of all invocations of `StateStore#flush` to instead call
`StateStore#commit`.

The `changelogOffsets` being passed in these invocations is currently
unused: the behaviour of `StateStore#commit` remains identical to
`StateStore#flush` before these changes.

A new implementation of `StateStore#commit` that actually uses these
offsets, along with changes to the use-site (in `ProcessorStateManager`
and `GlobalStateManager`) will come in a later PR.

Many strings, including documentation, and some variable names, have
also been renamed (from "flush" to "commit"), to maintain consistency
with the method they relate to.

One exception is the `flush-rate` metric, which has not been renamed,
because it will instead be deprecated in favour of a new `commit-rate`
metric, which will be introduced in another PR.

The only change in behaviour is as follows: calling `StateStore#flush`
from within a `Processor` is now a guaranteed no-op.

In the future, this will throw an `UnsupportedOperationException`, but
to ensure no changes to end-user experience, we currently make it a
no-op.

Previously, any call to `StateStore#flush` from a `Processor` would have
made no difference to program semantics, but likely would introduce
performance problems for RocksDB. This is because it would force a flush
of RocksDB memtables to disk on every invocation, which if naively used
could be on _every_ `Record`.

Consequently, making this a no-op should not make a difference for
end-users, except potentially improving performance if they were
incorrectly calling this method.

Reviewers: Eduwer Camacaro <eduwerc@gmail.com>, Bill Bejeck
 <bbejeck@apache.org>

nicktelford committed:
KAFKA-14711: Add new StateStore interfaces

These are the new interfaces detailed in KIP-1035: "StateStore managed
changelog offsets".

This PR introduces the interfaces changes, but makes otherwise no
consequential behavioural changes.

Outside of `StateStore.java`, _all_ changes are essentially just a
rename of all invocations of `StateStore#flush` to instead call
`StateStore#commit`.

The `changelogOffsets` being passed in these invocations is currently
unused: the behaviour of `StateStore#commit` remains identical to
`StateStore#flush` before these changes.

A new implementation of `StateStore#commit` that actually uses these
offsets, along with changes to the use-site (in `ProcessorStateManager`
and `GlobalStateManager`) will come in a later PR.

Many strings, including documentation, and some variable names, have
also been renamed (from "flush" to "commit"), to maintain consistency
with the method they relate to.

One exception is the `flush-rate` metric, which has not been renamed,
because it will instead be deprecated in favour of a new `commit-rate`
metric, which will be introduced in another PR.

---

The only change in behaviour is as follows: calling `StateStore#flush`
from within a `Processor` is now a guaranteed no-op.

In the future, this will throw an `UnsupportedOperationException`, but
to ensure no changes to end-user experience, we currently make it a
no-op.

Previously, any call to `StateStore#flush` from a `Processor` would have
made no difference to program semantics, but likely would introduce
performance problems for RocksDB. This is because it would force a flush
of RocksDB memtables to disk on every invocation, which if naively used
could be on _every_ `Record`.

Consequently, making this a no-op should not make a difference for
end-users, except potentially improving performance if they were
incorrectly calling this method.

bbejeck reviewed (commented):
Thanks for the PR @nicktelford - I made a pass and it overall looks good, since this has been around for some time, can you rebase?

eduwercamacaro reviewed (approved):
Looks good to me

nicktelford commented:
@bbejeck @eduwercamacaro Thanks for the reviews. Rebased as requested.

eduwercamacaro commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractReadWriteDecorator.java:
```diff
@@ -48,6 +50,11 @@ public void init(final StateStoreContext stateStoreContext,
     }
 
+    @Override
+    public void commit(final Map<TopicPartition, Long> changelogOffsets) {
```
@nicktelford I just noticed that this integration test is broken: `org.apache.kafka.streams.integration.StandbyTaskEOSIntegrationTest.shouldWipeOutStandbyStateDirectoryIfCheckpointIsMissing`

I think you need to override the `commit()` method on the `KeyValueStoreReadWriteDecorator` inner class and delegate the call to the wrapped store. That class is defined here on the line 81.

nicktelford replied:
Hmm, not sure how I missed this. I must have not run the full test suite properly!

The correct solution here is to remove the call to `store.commit(Map.of())` from the `Processor` used in the integration test, as `commit` should never be called from a user's `Processor`.

There looks to be a similar bug in `EosIntegrationTest` too. I'll fix them both and re-run the full integration test suite.

Thanks for spotting this!

nicktelford replied:
I see what's happened here. At some point, nearly all the integration tests were moved out into a separate `:streams:integration-tests` gradle module, and I didn't notice. I kept running `:streams:integrationTest` blissfully unaware that I wasn't actually running them :laughing:

nicktelford committed:
Fix failing integration tests

The `Processor` in each of these tests was incorrectly updated to call
`store.commit(Map.of())`.

This method cannot be called by user processors, because doing so would
put the system into an inconsistent state.

Removing these calls solves the problem, and has no impact on the
correct running of these tests, because the writes to these stores are
visible immediately to the current transaction anyway.

nicktelford committed:
Remove unused import

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/StateStore.java:
```diff
@@ -73,8 +76,55 @@ public interface StateStore {
+    /**
+     * Commit all written records to this StateStore.
+     * <p>
+     * This method <b>CANNOT<b> be called by users from {@link org.apache.kafka.streams.processor.api.Processor
```
```suggestion
     * This method <b>CANNOT</b> be called by users from {@link org.apache.kafka.streams.processor.api.Processor
```

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -484,36 +484,36 @@ void restore(final StateStoreMetadata storeMetadata, final List<ConsumerRecord<b
      * @throws TaskMigratedException recoverable error sending changelog records that would cause the task to be removed
-     * @throws StreamsException fatal error when flushing the state store, for example sending changelog records failed
-     *                          or flushing state store get IO errors; such error should cause the thread to die
+     * @throws StreamsException fatal error when commmitting the state store, for example sending changelog records failed
```
```suggestion
     * @throws StreamsException fatal error when committing the state store, for example sending changelog records failed
```

bbejeck reviewed (commented):
Thanks for the PR @nicktelford this LGTM just 2 nits

nicktelford committed:
Fix JavaDoc typo

Co-authored-by: Bill Bejeck <bbejeck@gmail.com>

nicktelford committed:
Fix JavaDoc typo

Co-authored-by: Bill Bejeck <bbejeck@gmail.com>

bbejeck reviewed (approved):
LGTM

bbejeck commented:
merged #20955 into trunk

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -951,7 +952,8 @@ public long approximateNumEntries(final DBAccessor accessor) throws RocksDBExcep
-        public void flush(final DBAccessor accessor) throws RocksDBException {
+        public void commit(final DBAccessor accessor,
+                           final Map<TopicPartition, Long> changelogOffsets) throws RocksDBException {
             accessor.flush(columnFamily);
```
Wondering why this `flush(columnFamily)` call is not updated to `commit(changelogOffsets)` or similar?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStoreTest.java:
```diff
@@ -224,10 +224,10 @@ public void shouldDelegateAndRecordMetricsOnGetWithTimestamp() {
 
-        verify(inner).flush();
+        verify(inner).commit(Map.of());
         assertThat((Double) getMetric("flush-rate").metricValue(), greaterThan(0.0));
```
@nicktelford @bbejeck -- just came across this code on some other PR review, and wondering if we need to update this, and also check if the new `commit-rate` metric was bumped?

Same question for other store types?

We are just adding a new metered store and corresponding test -- if this will be changed in a follow up PR, please make sure to also cover the newly added `MeteredTimestampedKeyValueStoreWithHeadersTest`

Cf https://github.com/apache/kafka/pull/21451
